dali/base/sysinfologger.cpp

```diff
@@ -0,0 +1,891 @@
+        if (!conn)
+            return false;
+        xpath.set("//" MSG_NODE);
+        if (filter->queryHiddenOnly())
+            xpath.append("[" ATTR_HIDDEN "='1')]");
+        if (filter->queryVisibleOnly())
+            xpath.append("[" ATTR_HIDDEN "='0')]");
+        if (!isEmptyString(filter->queryMatchSource()))
+            xpath.appendf("[" ATTR_SOURCE "='%s']", filter->queryMatchSource());
+        if (filter->queryMatchCode())
+            xpath.appendf("[" ATTR_CODE "='%d']", (int)filter->queryMatchCode());
+        if (filter->queryMatchAudience()!=MSGAUD_all)
+            xpath.appendf("[" ATTR_AUDIENCE "='%s']", LogMsgAudienceToFixString(filter->queryMatchAudience()));
+        if (filter->queryMatchClass()!=MSGCLS_all)
+            xpath.appendf("[" ATTR_CLASS "='%s']", LogMsgClassToFixString(filter->queryMatchClass()));
+        if (filter->queryMatchMsgId())
+            xpath.appendf("[" ATTR_MSGID "='%" I64F "u']", filter->queryMatchMsgId());
+        if (filter->queryMatchTimeStamp())
+            xpath.appendf("[" ATTR_TIMESTAMP "='%" I64F "u']", filter->queryMatchTimeStamp());
+        root.setown(conn->getRoot());
+        msgIter.setown(root->getElements(xpath.str()));
+        msgIter->first();
+        return ensureMatch();
+    }
+    virtual bool next() override
+    {
+        msgIter->next();
+        return ensureMatch();
+
+    }
+    virtual bool isValid() override
+    {
+        return msgIter ? msgIter->isValid() : false;
+    }
+};
+
+ISysInfoLoggerMsgIterator * createSysInfoLoggerMsgIterator(bool visibleOnly, bool hiddenOnly, unsigned year, unsigned month, unsigned day)
+{
+    Owned<CSysInfoLoggerMsgFilter> filter = new CSysInfoLoggerMsgFilter(visibleOnly, hiddenOnly, year, month, day);
+    return new CSysInfoLoggerMsgIterator(filter.getClear(), false);
+}
+
+ISysInfoLoggerMsgIterator * createSysInfoLoggerMsgIterator(unsigned __int64 msgId, unsigned __int64 ts)
+{
+    Owned<CSysInfoLoggerMsgFilter> filter = new CSysInfoLoggerMsgFilter(msgId, ts);
+    return new CSysInfoLoggerMsgIterator(filter.getClear(), false);
+}
+
+ISysInfoLoggerMsgIterator * createSysInfoLoggerMsgIterator(ISysInfoLoggerMsgFilter * msgFilter)
+{
+    return new CSysInfoLoggerMsgIterator(msgFilter, false);
+}
+
+void logSysInfoError(const LogMsgId msgId, const LogMsgCategory & cat, LogMsgCode code, const char *source, const char * msg, unsigned __int64 ts)
+{
+    if (ts==0)
+        ts = getTimeStampNowValue();
+
+    if (isEmptyString(source))
+        source = "unknown";
+
+    StringBuffer xpath(SYS_INFO_ROOT);
+    unsigned year, month, day;
+    CDateTime timeStamp;
+    timeStamp.setTimeStamp(ts);
+    timeStamp.getDate(year, month, day);
+    xpath.appendf("/m%04u%02u/d%02u", year, month, day);
+
+    Owned<IRemoteConnection> conn = querySDS().connect(xpath.str(), myProcessSession(), RTM_LOCK_WRITE|RTM_CREATE_QUERY, SDS_LOCK_TIMEOUT);
```

Comment:
Will have to be careful with other writers/deleters/readers overlapping with locks to different paths.
May need all to lock to root to ensure writers have exclusivity.